### PR TITLE
Reject bidirectional formatting characters

### PR DESCRIPTION
**Pull request by josevalim**

### CODE REVIEW

File: lib/elixir/src/elixir_tokenizer.hrl [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+-define(is_quote(S), (S =:= $" orelse S =:= $')).
+-define(is_sigil(S), (S =:= $/ orelse S =:= $< orelse S =:= $" orelse S =:= $' orelse
+                      S =:= $[ orelse S =:= $( orelse S =:= ${ orelse S =:= $|)).
+
+%% Spaces
+-define(is_horizontal_space(S), (S =:= $\s orelse S =:= $\t)).
+-define(is_vertical_space(S), (S =:= $\r orelse S =:= $\n)).
+-define(is_space(S), (?is_horizontal_space(S) orelse ?is_vertical_space(S))).
+
+%% Bidirectional control
```
Comment (josevalim) [inline lib/elixir/src/elixir_tokenizer.hrl:21]:
```suggestion
%% Bidirectional control
%% Retrieved from https://trojansource.codes/trojan-source.pdf
```

**Comment by fertapric** [comment]
What do you think about warning on homoglyphs/confusables?

https://www.unicode.org/Public/security/10.0.0/confusables.txt 
https://github.com/rust-lang/rust/blob/master/compiler/rustc_parse/src/lexer/unicode_chars.rs

**Comment by josevalim** [comment]
@fertapric yes, that's a separate problem. Ideally we will merge this with the existing StringTokenizer. We don't need to keep all confusables, only the confusables that are currently valid as identifiers. And we can probably do it all with a single pass.

File: lib/elixir/test/elixir/kernel/errors_test.exs [hunk 2]
```diff
@@ -816,6 +816,24 @@ defmodule Kernel.ErrorsTest do
...
+                      'foo. # This is a \u202A'
+
+    assert_eval_raise SyntaxError,
+                      ~r"nofile:1:12: invalid bidirectional formatting character in string: \\u202A. If you want to use such character, use it in its escaped \\u202A form instead",
+                      '"this is a \u202A"'
+
+    assert_eval_raise SyntaxError,
+                      ~r"nofile:1:13: invalid bidirectional formatting character in string: \\u202A. If you want to use such character, use it in its escaped \\u202A form instead",
+                      '"this is a \\\u202A"'
+  end
```
Comment (ericmj) [inline lib/elixir/test/elixir/kernel/errors_test.exs:835]: Do we need a test that it is accepted in escaped form?

**Review by ericmj — APPROVED** [review]

File: lib/elixir/test/elixir/kernel/errors_test.exs (hunk 2 above)
Comment (josevalim) [reply lib/elixir/test/elixir/kernel/errors_test.exs:835]: My answer would be no. We always accept any encoded form.

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**